lib/puppet/parser/resource.rb

```diff
@@ -213,7 +213,13 @@ def set_parameter(param, value = nil)
     tag(*param.value) if param.name == :tag
 
     # And store it in our parameter hash.
-    @parameters[param.name] = param
+    # PUP-10511 duplicate param/value for Sensitive
+    if param.value.is_a?(Puppet::Pops::Types::PSensitiveType::Sensitive)
+      @parameters[param.name] = param.dup
+      @parameters[param.name].value = param.value.dup
+    else
+      @parameters[param.name] = param
+    end
```

Comment:
Prior to puppet 6, we accidentally duplicated parameter values which is wasteful. We changed it to freeze strings, but found some types and providers were relying on the dup behavior. In PUP-7141, we restricted the optimization when running in "master" run mode. In keeping with that theme, I think we don't want to dup sensitive types, but instead change how we evaluate sensitive types when "finishing" the catalog. Something like:

```ruby
diff --git a/lib/puppet/parser/resource.rb b/lib/puppet/parser/resource.rb
index 1b00206981..ed4ead9f29 100644
--- a/lib/puppet/parser/resource.rb
+++ b/lib/puppet/parser/resource.rb
@@ -338,10 +338,11 @@ class Puppet::Parser::Resource < Puppet::Resource
   end

   def replace_sensitive_data
-    parameters.each_pair do |name, param|
+    parameters.keys.each do |name|
+      param = parameters[name]
       if param.value.is_a?(Puppet::Pops::Types::PSensitiveType::Sensitive)
         @sensitive_parameters << name
-        param.value = param.value.unwrap
+        parameters[name] = Puppet::Parser::Resource::Param.from_param(param)
       end
     end
   end
```

And then add a `from_param` copy constructor to `Puppet::Parser::Resource::Param`.